miner/proposal_builder.go

```diff
@@ -317,141 +366,174 @@ func (pb *ProposalBuilder) decideMeshHash(ctx context.Context, current types.Lay
 				pb.cfg.minActiveSetWeight,
-				weight,
+				pb.shared.active.weight,
 				pb.cfg.layerSize,
 				pb.cfg.layersPerEpoch,
 			)
 		} else {
 			if ballot.EpochData == nil {
-				return fmt.Errorf("atx %d created invalid first ballot", pb.session.atx)
-			}
-			hash := ballot.EpochData.ActiveSetHash
-			set, err := activesets.Get(pb.cdb, hash)
-			if err != nil {
-				return fmt.Errorf("get activeset %s: %w", hash.String(), err)
-			}
-			var weight uint64
-			for _, id := range set.Set {
-				atx, err := pb.cdb.GetAtxHeader(id)
-				if err != nil {
-					return err
-				}
-				weight += atx.GetWeight()
+				return fmt.Errorf("atx %d created invalid first ballot", ss.session.atx)
 			}
-			pb.session.ref = ballot.ID()
-			pb.session.active.set = set.Set
-			pb.session.active.weight = weight
-			pb.session.eligibilities.slots = ballot.EpochData.EligibilityCount
+			ss.session.ref = ballot.ID()
+			ss.session.beacon = ballot.EpochData.Beacon
+			ss.session.eligibilities.slots = ballot.EpochData.EligibilityCount
 		}
 	}
-	if pb.session.eligibilities.proofs == nil {
-		pb.session.eligibilities.proofs = calcEligibilityProofs(
-			pb.signer.VRFSigner(),
-			pb.session.epoch,
-			pb.session.beacon,
-			pb.session.nonce,
-			pb.session.eligibilities.slots,
+	if ss.session.eligibilities.proofs == nil {
+		ss.session.eligibilities.proofs = calcEligibilityProofs(
+			ss.signer.VRFSigner(),
+			ss.session.epoch,
+			ss.session.beacon,
+			ss.session.nonce,
+			ss.session.eligibilities.slots,
 			pb.cfg.layersPerEpoch,
 		)
-		pb.logger.With().Info("proposal eligibilities for an epoch", log.Inline(pb.session))
+		ss.log.With().Info("proposal eligibilities for an epoch", log.Inline(&ss.session))
 		events.EmitEligibilities(
-			pb.session.epoch,
-			pb.session.beacon,
-			pb.session.atx,
-			uint32(len(pb.session.active.set)),
-			pb.session.eligibilities.proofs,
+			ss.session.epoch,
+			ss.session.beacon,
+			ss.session.atx,
+			uint32(len(pb.shared.active.set)),
+			ss.session.eligibilities.proofs,
 		)
 	}
 	return nil
 }
 
 func (pb *ProposalBuilder) build(ctx context.Context, lid types.LayerID) error {
-	latency := latencyTracker{start: time.Now()}
-	if err := pb.initSessionData(ctx, lid); err != nil {
+	start := time.Now()
+	if err := pb.initSharedData(ctx, lid); err != nil {
 		return err
 	}
-	latency.data = time.Now()
 
-	if lid <= pb.session.prev {
-		return fmt.Errorf("layer %d was already built", lid)
+	pb.mu.Lock()
+	// don't accept registration in the middle of computing proposals
+	signers := maps.Values(pb.signers)
+	pb.mu.Unlock()
+
+	var eg errgroup.Group
+	eg.SetLimit(pb.cfg.workersLimit)
+	for _, ss := range signers {
+		ss := ss
+		ss.latency.start = start
+		eg.Go(func() error {
+			if err := pb.initSignerData(ctx, ss, lid); err != nil {
+				if errors.Is(err, errAtxNotAvailable) {
+					ss.log.With().Info("smesher doesn't have atx that targets this epoch",
+						log.Context(ctx), log.Uint32("epoch", ss.session.epoch.Uint32()),
```

Comment:
types.EpochID implements `Field()` already:
```suggestion
						log.Context(ctx), ss.session.epoch,
```